Document limitRequired and complexityEstimators flags (#291)
neo4j/docs-graphql: modules/ROOT/pages/security/avoid-unbounded-queries.adoc

= Avoid unbounded queries
:description: This section describes how to prevent unbounded queries from reaching the server that can lead to performance issues.

> Unbounded queries can cause performance issues and potentially crash your application by consuming excessive resources. The Neo4j GraphQL Library provides several mechanisms to help you avoid these problems.

Reviewer: ```suggestion
Unbounded queries can cause performance issues and potentially crash your application by consuming excessive resources.
The Neo4j GraphQL Library provides several mechanisms to help you avoid these problems.
```

== Require `limit` argument

Setting `limitRequired` to `true` in the feature settings makes the `limit` argument non-nullable in the generated schema, forcing all read operations to include a `limit` argument on all fields that return a list of items. 

> This protects against queries that return massive datasets across nested relationships.

Reviewer: ```suggestion
This protects the server from queries that return massive datasets across nested relationships.
```

[source, javascript, indent=0]
----
const neoSchema = new Neo4jGraphQL({
    typeDefs,
    features: {
        limitRequired: true
    }
});
----

When this feature is enabled:

* All queries that return arrays must include a `limit` argument

> * Queries without a `limit` will not be compliant with the GraphQL schema

Reviewer: ```suggestion
* Queries without a `limit` are not compliant with the GraphQL schema
```

* This applies to both top-level queries and nested relationship queries

Assuming the following type definitions:
[source, graphql, indent=0]
----
type Movie @node { 
    title: String 
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN) 
}

type Actor @node {
    name: String
}
----

* Valid query:
[source, graphql, indent=0]
----
query {
    movies(limit: 1) {
        title
        actors(limit: 1) {
            name
        }
    }
}
----

* Invalid query, `limit` argument not provided on `movies` field:
[source, graphql, indent=0]
----
query {
    # Error: `Field "movies" argument "limit" of type "Int!" is required, but it was not provided`
    movies {
        title
        actors(limit: 1) {
            name
        }
    }
}
----

* Invalid query, `limit` argument not provided on `actors` field:
[source, graphql, indent=0]
----
query {
    movies(limit: 1) {
        title
        # Error: `Field "actors" argument "limit" of type "Int!" is required, but it was not provided`
        actors {
            name
        }
    }
}
----

== Complexity estimators

> Setting `complexityEstimators` to `true` enables the assignment of a complexity "score" to each field in your schema. This feature creates complexity estimators that are compatible with the link:https://www.npmjs.com/package/graphql-query-complexity[graphql-query-complexity] library.

Reviewer: ```suggestion
Setting `complexityEstimators` to `true` enables the assignment of a complexity "score" to each field in your schema.
This feature creates complexity estimators that are compatible with the link:https://www.npmjs.com/package/graphql-query-complexity[graphql-query-complexity] library.
```

* Each field in your schema is assigned a complexity score
* You can use the `getComplexity` function from `graphql-query-complexity` to calculate the total complexity of incoming queries
* You can set maximum complexity thresholds to reject overly complex queries

=== Example usage

To use the complexity estimators, include this logic to run when receiving a query in your GraphQL server implementation.
For example, if you are using Apollo Server, you can use the `requestDidStart` function of an `ApolloServerPlugin`.

[source, javascript, indent=0]
----
const complexity = getComplexity({
   schema, // result of Neo4jGraphQL.getSchema() 
   query,  // the GraphQL query document sent to server for execution
   estimators: DefaultComplexityEstimators, // exported from the neo4j/graphql package
});

// Define your maximum complexity threshold
const complexityThreshold = 1000; 
if (complexity > complexityThreshold) {
    throw new Error(`Query is too complex: ${complexity}. Maximum allowed complexity is ${complexityThreshold}.`);
}
----

=== Computing the complexity

The provided complexity estimators compute the complexity of a query using the following formula:

> - Each field has a base complexity of 1

Reviewer: ```suggestion
* Each field has a base complexity of `1`
```

> - If a field returns a list, the complexity of its children is multiplied by the `limit` argument provided on that field to which the field's complexity is added

Reviewer: ```suggestion
* If a field returns a list, its complexity is calculated by first adding the complexity values of its children and then multiplying it with the `limit` argument provided on that field (see examples below)
```

> - If no `limit` argument is provided, a default multiplier of 1 is used

Reviewer: ```suggestion
* If no `limit` argument is provided, a default multiplier of `1` is used
```

[NOTE]
====
c_query = c_child * lvl_limit + c_field 

where c_field = 1, lvl_limit defaults to 1

> ====

Reviewer: not sure the formula captures all of it -
there's no sum over the children, is there?

also, let's not include it in a note, but rather as part of the normal text

> The queries below demonstrate the usage of the formula.

Reviewer: maybe we could add a hint that the calculation is to be read bottom-up (took me a second to figure that out at least :) )

= Avoid unbounded queries
:description: This section describes how to prevent unbounded queries from reaching the server that can lead to performance issues.

Unbounded queries can cause performance issues and potentially crash your application by consuming excessive resources.
The Neo4j GraphQL Library provides several mechanisms to help you avoid these problems.

== `@limit` directive

> Using the xref::/directives/custom-logic.adoc#_limit[`@limit`] directive on a type definition allows to set a default and a maximum limit on the number of nodes returned for that type from the server side. This is useful to prevent clients from accidentally or intentionally requesting too much data in a single query.

Reviewer: ```suggestion
Using the xref::/directives/custom-logic.adoc#_limit[`@limit`] directive on a type definition allows to set a default value and a maximum for the limit on the number of nodes returned for that type from the server side.
This is useful to prevent clients from accidentally or intentionally requesting large amounts of data in a single query.
```

> == Require `limit` argument

Reviewer 2: Is this a subheader of `@limit` directive? Shouldn't it be `===`?

Author: no, it's a different thing, it's the requireLimit constructor flag

      edges {                                                           # 12 + 2x
        node {                                                          # 9 + 2x + 1 + 1 = 11 + 2x
          title                                                         # 1              
          actorsConnection(first: 2, sort: { node: { name: ASC } }) {   # (4 + x) * 2 + 1 = 9 + 2x
            edges {                                                     # 3 + x + 1 = 4 + x
              node {                                                    # 2
                name                                                    # 1
              }
              properties {                                              # 1 + x; x = max nr of properties in below fragments
                ... on ActedIn {
                  roles
                }
                ... on ActedInSeries {
                  roles
                  episodes
                  year
                }
              }
            }
          }
        }
      }
  }
}
# Total complexity = 1 + 10 * (1 + 1 + 1 + 1 + 2 * (1 + 1 + 1 + 1 + 3)) = 181
----
====

=== Complexity considerations

> It is intended for this feature to be used in conjunction with `limitRequired` to provide the most effective protection against unbounded queries. 

Reviewer: ```suggestion
This feature is intended to be used in conjunction with xref::neo4jgraphql-class.adoc#_limit_required_settings[`limitRequired`] to provide the most effective protection against unbounded queries. 
```

> When `limitRequired` is enabled, the complexity score of a query will be higher due to the multiplication by the provided `limit` arguments, which more accurately reflects the potential cost of executing that query.

Reviewer: ```suggestion
When `limitRequired` is enabled, the complexity score of a query is higher because the provided `limit` arguments act as multipliers, which more accurately reflects the potential cost of executing that query.
```

Reviewer: this one has not been applied -
does the wording suggestion not capture the intended meaning?

Author: Hmmm I think it doesn't. It's not the score that's multiplied, the score is a result. It's the equation generating that result that involves multiplication with the limit arguments.

Reviewer: ok, updated the suggestion
still wondering though - what could we call the things that get multiplied with the limit arguments? intermediate results or something like that?

Author: or maybe sth like "the provided `limit` arguments are multipliers"? And then we don't need a way to refer the intermediate results?

Reviewer: updated the wording

> When `limitRequired` is not enabled, the complexity score will be orders of magnitude lower since it uses a default multiplier of 1 for fields that return lists. 

Reviewer: ```suggestion
When `limitRequired` is not enabled, the complexity score is orders of magnitude lower since it uses a default multiplier of 1 for fields that return lists.
```

> While it still provides a useful measure of query complexity based on the structure of the query and the number of fields requested, it will be less effective at preventing unbounded queries and the complexity threshold will be difficult to set (Examples 1 and 2 above demonstrate a major difference in complexity scores: 5 and 1021).

Reviewer: ```suggestion
While it still provides a useful measure of query complexity based on the structure of the query and the number of fields requested, it is less effective at preventing unbounded queries and the complexity threshold will be difficult to set.
For instance, the first two examples above demonstrate a major difference in complexity scores: 5 versus 1021.
```

=== Alternative solutions

> An alternative to compute the query complexity is to use a tool such as link:https://escape.tech/graphql-armor/docs/plugins/cost-limit/[GraphQL Armor].

Reviewer: ```suggestion
An alternative to computing the query complexity is to use a tool such as link:https://escape.tech/graphql-armor/docs/plugins/cost-limit/[GraphQL Armor].
```
